File: app/src/main/java/com/github/malitsplus/shizurunotes/ui/charadetails/CharaDetailsViewModel.java
package com.github.malitsplus.shizurunotes.ui.charadetails;

import android.app.Application;

import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.ViewModel;

import com.github.malitsplus.shizurunotes.data.Chara;
import com.github.malitsplus.shizurunotes.ui.SharedViewModel;

public class CharaDetailsViewModel extends ViewModel {

    private SharedViewModel sharedViewModel;
    public MutableLiveData<Chara> mutableChara = new MutableLiveData<>();

    public CharaDetailsViewModel(SharedViewModel sharedViewModel){
        this.sharedViewModel = sharedViewModel;
        setChara(sharedViewModel.getSelectedChara());
    }

    public void setChara(Chara chara){
        this.mutableChara.setValue(chara);
    }

    public LiveData<Chara> getMutableChara(){
        return mutableChara;
    }
}

File: app/src/main/java/com/github/malitsplus/shizurunotes/ui/charadetails/CharaDetailsViewModel.kt
package com.github.malitsplus.shizurunotes.ui.charadetails

import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.ViewModel
import com.github.malitsplus.shizurunotes.data.Chara
import com.github.malitsplus.shizurunotes.ui.shared.SharedViewModelChara

class CharaDetailsViewModel(private val sharedViewModelChara: SharedViewModelChara) : ViewModel() {

    val mutableChara = MutableLiveData<Chara>()

    fun changeRank(rankString: String){
        val rank = rankString.toInt()
        val chara = mutableChara.value?.shallowCopy()
        chara?.apply {
            setCharaProperty(rank = rank)
            skills.forEach {
                it.setActionDescriptions(maxCharaLevel, charaProperty)
            }
        }
        mutableChara.value = chara
    }

    fun setChara(chara: Chara?) {
        mutableChara.value = chara
    }

    fun getChara(): Chara?{
        return mutableChara.value
    }

    fun getAttackPatternList(): List<Any> {
        val list = mutableListOf<Any>()
        mutableChara.value?.let { chara ->
            for (i in 1..chara.attackPatternList.size) {
                list.add(i)
                chara.attackPatternList[i - 1].items.forEach {
                    list.add(it)
                }
            }
        }
        return list
    }

    init {
        setChara(sharedViewModelChara.selectedChara)
    }
}